includes/class-llms-media-protector.php

```diff
@@ -0,0 +1,640 @@
+	 *
+	 * @todo Maybe respond with 404 Not Found?
+	 * @todo Maybe respond with 401 Unauthorized?
+	 *
+	 * @since [version]
+	 *
+	 * @return void
+	 */
+	public function serve_file() {
+
+		$media_id = $_GET[ self::QUERY_PARAMETER_ID ];
+		$media_file = get_post( $media_id );
+
+
+		// Validate that the attachment post exists.
+		if ( is_null( $media_file ) ) {
+			return;
+		}
+
+		$file_name = $this->get_media_path( $media_id );
+
+		$size = null;
+		if ( isset( $_GET[ self::QUERY_PARAMETER_SIZE ] ) ) {
+			if ( '[' === $_GET[ self::QUERY_PARAMETER_SIZE ][0] ) {
+				$size = json_decode( $_GET[ self::QUERY_PARAMETER_SIZE ] );
+			} else {
+				$size = $_GET[ self::QUERY_PARAMETER_SIZE ];
+			}
+		}
+
+		$icon = null;
+		if ( isset( $_GET[ self::QUERY_PARAMETER_ICON ] ) ) {
+			$icon = (bool) $_GET[ self::QUERY_PARAMETER_ICON ];
+		}
+
+		// Optionally, use an alternate image size.
+		if ( ! is_null( $size ) || ! is_null( $icon ) ) {
+			$image     = wp_get_attachment_image_src( $media_id, $size, $icon );
+			$base_name = basename( $image[0] );
+			$file_name = dirname( $file_name ) . "/$base_name";
+		}
+
+		// Validate that the media file exists.
+		if ( false === file_exists( $file_name ) ) {
+			return;
+		}
+
+		// Is the user authorized to view the file?
+		$is_authorized = $this->is_authorized_to_view( get_current_user_id(), $media_id );
+		if ( false === $is_authorized ) {
+			$content_type = $media_file->post_mime_type;
+			if ( 0 === stripos( $content_type, 'image/' ) ) {
+				// @todo Find or create an image to denote unauthorized access to an image file.
+				$file_name = LLMS_PLUGIN_DIR . 'assets/images/circle-with-a-slash.png';
+			} else {
+				header( 'HTTP/1.1 403 Forbidden' );
+				exit();
+			}
+		}
+
+		/**
+		 * Determine how the media file should be served.
+		 *
+		 * @since [version]
+		 *
+		 * @param string $serve_method One of the {@see LLMS_Media_Protector::SERVE_REDIRECT LLMS_Media_Protector::SERVE_X} constants.
+		 * @param int    $media_id     The post ID of the media file.
+		 */
+		$serve_method = apply_filters( 'llms_media_serve_method', self::SERVE_SEND_FILE, $media_id );
+
+		switch ( $serve_method ) {
+			case self::SERVE_READ_FILE:
+				$this->send_headers( $file_name, $media_id );
+				$this->read_file( $file_name );
+				break;
+			case self::SERVE_SEND_FILE:
+				$this->send_file( $file_name, $media_id );
+				break;
+			case self::SERVE_REDIRECT:
+			default:
+				$this->send_redirect( $media_id );
+				break;
+		}
+	}
+
+	/**
+	 * Sets the upload subdirectory.
+	 *
+	 * @since [version]
+	 *
+	 * @param string $upload_subdirectory
+	 * @return $this
+	 */
+	public function set_upload_subdirectory( string $upload_subdirectory ) {
+
+		// Add leading slash.
+		if ( strpos( $upload_subdirectory, '/' ) !== 0 ) {
+			$upload_subdirectory = '/' . $upload_subdirectory;
+		}
+
+		// Strip trailing slash.
+		$upload_subdirectory = rtrim( $upload_subdirectory, '/' );
```

Comment:
Nitpicking but I like using the WP api where possible: https://developer.wordpress.org/reference/functions/untrailingslashit/

```suggestion
		$upload_subdirectory = untrailingslashit( $upload_subdirectory );
```